### DISCUSSION TITLE
Program can't find netcdf file

### DISCUSSION THREAD

**Discussion by darylsuen** [neuralhydrology/neuralhydrology#183] [General]
I'm trying to run a very simple generic dataset. The dataset file name is '1.nc' and is in the 'time_series' folder. There is a txt file in the 'BasinIntFiles' folder that has the integer '1' on the first line. I'm running nh-run train from a cmd window. The program seems to be reading the txt file correctly and searches for a netcdf file named '1' in the 'time_series' folder but can't find it. The error message is:

    raise FileNotFoundError(f"No netCDF file found for basin {basin} in {files_dir}")
FileNotFoundError: No netCDF file found for basin ï»¿1 in C:\Users\daryl\OneDrive\Documents\GDAA3000\ProjectDischarge\RdrsSample\NhDataRootDir\basin_mean_forcing\NovaScotiaRDRS\time_series

Any ideas?

**Comment by darylsuen** [comment]
I simplified my file structure a little bit and ran the code again. I'm using a different basin and its netcdf file is named '2.nc'.  The program is looking in the correct folder for the netcdf file. The error is: 

FileNotFoundError: No netCDF file found for basin ﻿2 in C:\Users\daryl\OneDrive\Documents\GDAA3000\ProjectDischarge\RdrsSample\NhDataRootDir\basin_mean_forcing\time_series

Below, is a screenshot of the 'basin_mean_forcing' directory and you can see part of the full file path in the address bar. The full file path copied and pasted from the address bar is: C:\Users\daryl\OneDrive\Documents\GDAA3000\ProjectDischarge\RdrsSample\NhDataRootDir\basin_mean_forcing

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/bcb3c036-3a87-4553-b1d5-cbb9b10f64ab)

You can see the '2.nc' file in the 'time_series' folder.

Copied and pasted from the 'data_dir:' setting in the .yml file: 

data_dir: C:\Users\daryl\OneDrive\Documents\GDAA3000\ProjectDischarge\RdrsSample\NhDataRootDir\basin_mean_forcing

I looked at the 'genericdataset.py' file and the relevant code in the 'load_timeseries()' function and I don't see what could be causing the issue. The code is:

    files_dir = data_dir / "time_series"
    netcdf_files = list(files_dir.glob("*.nc4"))
    netcdf_files.extend(files_dir.glob("*.nc"))
    netcdf_file = [f for f in netcdf_files if f.stem == basin]
    if len(netcdf_file) == 0:
        raise FileNotFoundError(f"No netCDF file found for basin {basin} in {files_dir}")
    if len(netcdf_file) > 1:
        raise ValueError(f"Multiple netCDF files found for basin {basin} in {files_dir}")

Here is a screen shot of the code because it is easier to read with the colors:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/20272215-c77f-4b58-a312-e95ba32aac64)

**Reply by gauchm** [reply]
Just a guess, but maybe the integer names are throwing it off? Have you tried renaming that basin to some string name, like "basin_2"?

**Reply by darylsuen** [reply]
Yes, the original name was 01DP001. However, the names used by CAMELS US are integers so I thought, at first, that my original name was throwing off the system. Also, the code in genericdataset.py shown in the image doesn't ask for a specific data type.  the files_dir.glob("*.nc") just gets the paths ending with '.nc' , so as long as it's a path it add is to the list of netcdf_files. 

I think the problem lies in this line of code from the genericdataset.py:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/23ebc32e-78e6-4a83-8b28-f057725e9da9)

The stem of the path, which is the filename without the extension, must be the same as the value assigned to 'basin'. If it isn't, it will not be added to the netcdf_file list and then len(netcdf_file) ==0: returns True, causing the raise ValueError.

The code is a component of the load_timeseries() of the genericdataset class. This function takes 'data_dir' and 'basin' as arguments, so I suspect that there is a setting in the config file for 'basin' that does not have the name of the file, in this case '1' as its value.

**Reply by darylsuen** [reply]
The problem was that 'basin' has a non-printable character in front of it the value that isn't seen in the code. I had to edit the genericdataset.py file in the datasetzoo directory to remove the character like this:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/a61fd7bd-0076-46bf-89dd-d391faa7b6b8)

You'll notice that I also added some lines to the error message so you can see what is in the lists. The program will run after making this edit, if this is your problem.
